pkg/snapshotmgr/snapshot_manager.go

```diff
@@ -244,65 +253,98 @@ func (this *SnapshotManager) createSnapshot(peID astrolabe.ProtectedEntityID, ta
 		return updatedPeID, nil
 	}
 
+	_, err = this.UploadSnapshot(peID, updatedPeID, peSnapID, pe, ctx, backupRepositoryName, snapshotRef)
+	if err != nil {
+		return astrolabe.ProtectedEntityID{}, err
+	}
+
+	return updatedPeID, nil
+}
+
+/*
+Creates an Upload CR
+*/
+func (this *SnapshotManager) UploadSnapshot(peID astrolabe.ProtectedEntityID, updatedPeID astrolabe.ProtectedEntityID,
+	peSnapID astrolabe.ProtectedEntitySnapshotID, pe astrolabe.ProtectedEntity, ctx context.Context, backupRepositoryName string,
+	snapshotRef string) (*v1api.Upload, error) {
 	this.Info("Start creating Upload CR")
 	config, err := rest.InClusterConfig()
 	if err != nil {
 		this.WithError(err).Errorf("Failed to get k8s inClusterConfig")
-		return updatedPeID, err
+		return nil, err
 	}
 	pluginClient, err := plugin_clientset.NewForConfig(config)
 	if err != nil {
 		this.WithError(err).Errorf("Failed to get k8s clientset from the given config: %v ", config)
-		return updatedPeID, err
+		return nil, err
 	}
 
 	// look up velero namespace from the env variable in container
 	veleroNs, exist := os.LookupEnv("VELERO_NAMESPACE")
 	if !exist {
 		this.WithError(err).Errorf("CreateSnapshot: Failed to lookup the env variable for velero namespace")
-		return updatedPeID, err
+		return nil, err
 	}
 
-	uploadBuilder := builder.ForUpload(veleroNs, "upload-"+peSnapID.GetID()).
+	uploadUUID, err := uuid.NewRandom()
+	if err != nil {
+		this.WithError(err).Errorf("Could not create upload UUID")
+		return nil, err
+	}
+	uploadName := "upload-" + uploadUUID.String()
+	this.Info("Creating Upload CR: %s/%s", veleroNs, uploadName)
+	uploadBuilder := builder.ForUpload(veleroNs, uploadName).
 		BackupTimestamp(time.Now()).
 		NextRetryTimestamp(time.Now()).
 		Phase(v1api.UploadPhaseNew).
 		SnapshotReference(snapshotRef)
-
 	if peID.GetPeType() == astrolabe.PvcPEType {
 		components, err := pe.GetComponents(ctx)
 		if err != nil {
 			this.WithError(err).Errorf("Failed to retrive subcomponents for %s", peID.String())
-			return updatedPeID, err
+			return nil, err
 		}
 		if len(components) != 1 {
-			return updatedPeID, errors.New(fmt.Sprintf("Expected 1 component, %s has %d", peID.String(), len(components)))
+			return nil, errors.New(fmt.Sprintf("Expected 1 component, %s has %d", peID.String(), len(components)))
+		}
+
+		// Extract Snapshot ID
+		componentID64Str := peSnapID.String()
```

Comment:
I am not understanding this part. The components are extracted at line 302, they were previously used to generate the component pe-id which became a snap ref in the upload CRD. The current change uses the peSnapId directly. Also why the decodeString?